File: python/keel/src/keel/_sim.py
"""`keel sim`: adapter-level fault injection driven by a declarative fault
plan (`KEEL_SIM_PLAN=<path>`, set by `keel sim <plan>` on the child).

See ``docs/sim-format.md`` for the full (non-contract) plan format and the
architecture-spec §8 rationale. In short: this module wraps the ADAPTER's
``effect`` closure — the callable ``Backend.execute``/``execute_async`` calls
once per Tier 1 attempt — never ``execute``'s return value, so a scripted
failure is genuinely retried/backed-off/breaker-tripped by the real backend's
own resilience logic, exactly like a real failure would be. Every other
``Backend`` member (``enter_flow``/``exit_flow``/``journal_time``/
``journal_random``/``report``/``layer``) delegates straight through
(``__getattr__``, mirroring ``_backend._NativeBackend`` and
``_record.RecordingBackend``) — Tier 2 flow control and journal semantics are
completely untouched by fault injection.
"""

from __future__ import annotations

import asyncio
import json
import os
import signal
import sys
import time
from pathlib import Path
from typing import Any, Callable, Mapping

SIM_VERSION = 1

#: 128 + SIGKILL(9) — the exit code a POSIX shell reports for a process a
#: real `kill -9` terminated. Used as the fallback hard-exit code on a
#: platform with no SIGKILL (documented in docs/sim-format.md).
SIM_CRASH_EXIT_CODE = 137

#: Per-`kind` default HTTP status when the directive does not name one.
_DEFAULT_STATUS = {"5xx": 503, "429": 429, "http": 500}

_TRUTHY = {"1", "true", "yes"}


def _default_crash() -> None:
    """Hard-crash this process right now: no cleanup, no atexit, no exception
    unwinding — the closest in-process model of a real `kill -9` (module
    docs). A real, uncatchable `SIGKILL` sent to our own pid mirrors the
    docs' own vocabulary exactly; `os._exit` is the fallback on a platform
    with no SIGKILL (e.g. native Windows)."""
    try:
        os.kill(os.getpid(), signal.SIGKILL)
    except (AttributeError, OSError):
        pass
    os._exit(SIM_CRASH_EXIT_CODE)  # pragma: no cover - SIGKILL never returns on POSIX


class _Cursor:
    """Per-target consumed-directive counters, persisted to a JSON sidecar
    next to the plan file (`<plan path>.cursor.json`) so a `"crash"`
    directive's hard restart resumes the SAME fault sequence instead of
    replaying it from the top — mirrors how the flow journal itself survives
    a real crash. Best-effort: a read/write failure degrades to an in-memory
    (non-persisted) cursor rather than breaking the simulated program."""

    def __init__(self, path: Path) -> None:
        self._path = path
        try:
            loaded = json.loads(path.read_text())
            self._counts: dict[str, int] = loaded if isinstance(loaded, dict) else {}
        except (OSError, ValueError):
            self._counts = {}

    def next_index(self, target: str) -> int:
        return self._counts.get(target, 0)

    def bump(self, target: str) -> None:
        self._counts[target] = self.next_index(target) + 1
        try:
            self._path.write_text(json.dumps(self._counts))
            # fsync so the counter survives the crash we may be about to
            # trigger (best-effort: some filesystems/paths cannot be opened
            # read-only for fsync on every platform — never fatal).
            fd = os.open(self._path, os.O_RDONLY)
            try:
                os.fsync(fd)
            finally:
                os.close(fd)
        except OSError:
            pass


def _directive_at(directives: list[dict[str, Any]], index: int) -> dict[str, Any] | None:
    """The directive `index` (0-based, across every attempt this target has
    ever seen) selects, honoring each entry's `repeat` (default 1) — mirrors
    `tools/faultproxy`'s `Scenario` cursor. `None` once the sequence is spent
    (every further attempt passes through to the real effect live)."""
    remaining = index
    for directive in directives:
        span = max(1, int(directive.get("repeat", 1)))
        if remaining < span:
            return directive
        remaining -= span
    return None


def _resolve(directive: dict[str, Any]) -> dict[str, Any] | None | str:
    """`None` → passthrough (call the real effect). `"crash"` → the caller
    must hard-crash. Otherwise the synthetic `AttemptResult` dict to return
    without ever calling the real effect."""
    kind = directive.get("kind", "ok")
    if kind == "ok":
        return None
    if kind == "crash":
        return "crash"
    if kind == "conn":
        return {"status": "error", "class": "conn", "message": "keel sim: injected connection failure"}
    if kind == "timeout":
        return {"status": "error", "class": "timeout", "message": "keel sim: injected timeout"}
    status = int(directive.get("status", _DEFAULT_STATUS.get(kind, 500)))
    result: dict[str, Any] = {
        "status": "error",
        "class": "http",
        "http_status": status,
        "message": f"keel sim: injected HTTP {status}",
    }
    if "retry_after_ms" in directive:
        result["retry_after_ms"] = directive["retry_after_ms"]
    return result


class SimBackend:
    """Wraps ``inner`` (a ``Backend``): for every ``execute``/``execute_async``
    call, wraps the caller's ``effect`` closure so a scripted fault plan can
    inject a failure/latency/crash into one Tier 1 attempt without ``inner``
    ever seeing anything but a normal (possibly synthetic) attempt outcome —
    its own retry/backoff/breaker/cache decisions run for real over it."""

    def __init__(
        self,
        inner: Any,
        faults: Mapping[str, list[dict[str, Any]]],
        cursor: _Cursor,
        crash: Callable[[], None] = _default_crash,
    ) -> None:
        self._inner = inner
        self._faults = faults
        self._cursor = cursor
        self._crash = crash

    def configure(self, policy: dict[str, Any]) -> None:
        self._inner.configure(policy)

    def _directives_for(self, request: Any) -> list[dict[str, Any]] | None:
        target = str(request.get("target", "")) if isinstance(request, dict) else ""
        return self._faults.get(target) or None

    def execute(self, request: dict[str, Any], effect: Callable[[int], dict[str, Any]]) -> dict[str, Any]:
        directives = self._directives_for(request)
        if directives is None:
            return self._inner.execute(request, effect)
        target = str(request.get("target", ""))

        def wrapped(attempt: int) -> dict[str, Any]:
            return self._apply(target, directives, attempt, effect)

        return self._inner.execute(request, wrapped)

    def report(self) -> dict[str, Any]:
        return self._inner.report()

    def _apply(
        self,
        target: str,
        directives: list[dict[str, Any]],
        attempt: int,
        effect: Callable[[int], dict[str, Any]],
    ) -> dict[str, Any]:
        index = self._cursor.next_index(target)
        directive = _directive_at(directives, index)
        self._cursor.bump(target)
        if directive is None:
            return effect(attempt)
        delay_ms = directive.get("delay_ms")
        if delay_ms:
            time.sleep(max(0, int(delay_ms)) / 1000.0)
        outcome = _resolve(directive)
        if outcome is None:
            return effect(attempt)
        if outcome == "crash":
            self._crash()
            raise AssertionError("unreachable: crash() must not return")  # pragma: no cover
        return outcome

    async def _apply_async(
        self,
        target: str,
        directives: list[dict[str, Any]],
        attempt: int,
        effect: Callable[[int], Any],
    ) -> dict[str, Any]:
        index = self._cursor.next_index(target)
        directive = _directive_at(directives, index)
        self._cursor.bump(target)
        if directive is None:
            return await effect(attempt)
        delay_ms = directive.get("delay_ms")
        if delay_ms:
            await asyncio.sleep(max(0, int(delay_ms)) / 1000.0)
        outcome = _resolve(directive)
        if outcome is None:
            return await effect(attempt)
        if outcome == "crash":
            self._crash()
            raise AssertionError("unreachable: crash() must not return")  # pragma: no cover
        return outcome

    def __getattr__(self, name: str) -> Any:
        attr = getattr(self._inner, name)
        if name == "execute_async" and callable(attr):

            async def _wrapped(request: dict[str, Any], effect: Callable[[int], Any]) -> dict[str, Any]:
                directives = self._directives_for(request)
                if directives is None:
                    return await attr(request, effect)
                target = str(request.get("target", ""))

                async def wrapped(a: int) -> dict[str, Any]:
                    return await self._apply_async(target, directives, a, effect)

                return await attr(request, wrapped)

            return _wrapped
        return attr


def _load_faults(plan_path: str) -> dict[str, list[dict[str, Any]]]:
    try:
        data = json.loads(Path(plan_path).read_text())
    except (OSError, ValueError) as exc:
        sys.stderr.write(f"keel ▸ KEEL_SIM_PLAN={plan_path!r} could not be read: {exc}\n")
        raise SystemExit(1) from exc
    faults = data.get("faults") if isinstance(data, dict) else None
    if not isinstance(faults, dict):
        return {}
    out: dict[str, list[dict[str, Any]]] = {}
    for target, directives in faults.items():
        if isinstance(target, str) and isinstance(directives, list):
            out[target] = [d for d in directives if isinstance(d, dict)]
    return out


def install_sim(
    backend: Any,
    *,
    plan_path: str,
    env: Mapping[str, str],
) -> SimBackend:
    """Wrap `backend` for `keel sim <plan>` (`KEEL_SIM_PLAN=<plan_path>`)."""
    faults = _load_faults(plan_path)
    cursor = _Cursor(Path(f"{plan_path}.cursor.json"))
    if env.get("KEEL_QUIET", "").strip().lower() not in _TRUTHY:
        sys.stderr.write(f"keel ▸ fault-simulating with {plan_path} — see docs/sim-format.md\n")
    return SimBackend(backend, faults, cursor)


__all__ = [
    "SIM_CRASH_EXIT_CODE",
    "SIM_VERSION",
    "SimBackend",
    "install_sim",
]

File: node/keel/src/sim.mjs
/**
 * `keel sim`: adapter-level fault injection driven by a declarative fault
 * plan (`KEEL_SIM_PLAN=<path>`, set by `keel sim <plan>` on the child).
 *
 * See `docs/sim-format.md` for the full (non-contract) plan format and the
 * architecture-spec §8 rationale. In short: this module wraps the ADAPTER's
 * `effect` closure — the callable `Backend.execute` calls once per Tier 1
 * attempt — never `execute`'s return value, so a scripted failure is
 * genuinely retried/backed-off/breaker-tripped by the real backend's own
 * resilience logic, exactly like a real failure would be. Every other
 * `Backend` member (`report`, `layer`, `persistent`) delegates straight
 * through, mirroring `record.mjs`'s `RecordingBackend`.
 */

import { readFileSync, writeFileSync, openSync, fsyncSync, closeSync } from "node:fs";

export const SIM_VERSION = 1;

/** 128 + SIGKILL(9) — the exit code a POSIX shell reports for a process a
 * real `kill -9` terminated (documented fallback on a platform with no
 * SIGKILL). */
export const SIM_CRASH_EXIT_CODE = 137;

/** Per-`kind` default HTTP status when the directive does not name one. */
const DEFAULT_STATUS = { "5xx": 503, "429": 429, http: 500 };

const TRUTHY = new Set(["1", "true", "yes"]);
function isTruthy(v) {
  return TRUTHY.has(String(v ?? "").trim().toLowerCase());
}

/** Hard-crash this process right now: a real, uncatchable `SIGKILL` sent to
 * our own pid — the closest in-process model of a real `kill -9` (module
 * docs). `process.exit` is deliberately NOT used: it runs `exit`/`beforeExit`
 * listeners (`installExitFlush`), which the simulation is specifically
 * trying to skip. */
export function defaultCrash() {
  try {
    process.kill(process.pid, "SIGKILL");
  } catch {
    /* no SIGKILL on this platform (native Windows) — fall through */
  }
  process.reallyExit ? process.reallyExit(SIM_CRASH_EXIT_CODE) : process.exit(SIM_CRASH_EXIT_CODE);
}

/**
 * Per-target consumed-directive counters, persisted to a JSON sidecar next
 * to the plan file (`<plan path>.cursor.json`) so a `"crash"` directive's
 * hard restart resumes the SAME fault sequence instead of replaying it from
 * the top. Best-effort: a read/write failure degrades to an in-memory
 * (non-persisted) cursor rather than breaking the simulated program.
 */
class Cursor {
  #path;
  #counts;
  constructor(path) {
    this.#path = path;
    try {
      const loaded = JSON.parse(readFileSync(path, "utf8"));
      this.#counts = loaded && typeof loaded === "object" && !Array.isArray(loaded) ? loaded : {};
    } catch {
      this.#counts = {};
    }
  }
  nextIndex(target) {
    return this.#counts[target] ?? 0;
  }
  bump(target) {
    this.#counts[target] = this.nextIndex(target) + 1;
    try {
      writeFileSync(this.#path, JSON.stringify(this.#counts));
      // fsync so the counter survives the crash we may be about to trigger.
      const fd = openSync(this.#path, "r");
      try {
        fsyncSync(fd);
      } finally {
        closeSync(fd);
      }
    } catch {
      /* best-effort bookkeeping only */
    }
  }
}

/** The directive `index` (0-based, across every attempt this target has ever
 * seen) selects, honoring each entry's `repeat` (default 1) — mirrors
 * `tools/faultproxy`'s Scenario cursor. `null` once the sequence is spent
 * (every further attempt passes through to the real effect live). */
function directiveAt(directives, index) {
  let remaining = index;
  for (const directive of directives) {
    const span = Math.max(1, Number(directive.repeat ?? 1));
    if (remaining < span) return directive;
    remaining -= span;
  }
  return null;
}

/** `null` → passthrough (call the real effect). `"crash"` → the caller must
 * hard-crash. Otherwise the synthetic attempt-result object to return
 * without ever calling the real effect. */
function resolve(directive) {
  const kind = directive.kind ?? "ok";
  if (kind === "ok") return null;
  if (kind === "crash") return "crash";
  if (kind === "conn") return { status: "error", class: "conn", message: "keel sim: injected connection failure" };
  if (kind === "timeout") return { status: "error", class: "timeout", message: "keel sim: injected timeout" };
  const status = Number(directive.status ?? DEFAULT_STATUS[kind] ?? 500);
  const result = { status: "error", class: "http", http_status: status, message: `keel sim: injected HTTP ${status}` };
  if ("retry_after_ms" in directive) result.retry_after_ms = directive.retry_after_ms;
  return result;
}

async function sleep(ms) {
  await new Promise((r) => setTimeout(r, ms));
}

/**
 * Wraps `inner` (a `Backend`): for every `execute` call, wraps the caller's
 * `effect` closure so a scripted fault plan can inject a failure/latency/
 * crash into one Tier 1 attempt without `inner` ever seeing anything but a
 * normal (possibly synthetic) attempt outcome — its own retry/backoff/
 * breaker/cache decisions run for real over it.
 */
export class SimBackend {
  #inner;
  #faults;
  #cursor;
  #crash;
  constructor(inner, faults, cursor, crash = defaultCrash) {
    this.#inner = inner;
    this.#faults = faults;
    this.#cursor = cursor;
    this.#crash = crash;
  }
  configure(policy) {
    return this.#inner.configure(policy);
  }
  #directivesFor(request) {
    const target = request && typeof request === "object" ? String(request.target ?? "") : "";
    return this.#faults[target] ?? null;
  }
  async execute(request, effect) {
    const directives = this.#directivesFor(request);
    if (directives === null) return this.#inner.execute(request, effect);
    const target = String(request.target ?? "");
    const wrapped = (attempt) => this.#apply(target, directives, attempt, effect);
    return this.#inner.execute(request, wrapped);
  }
  report() {
    return this.#inner.report();
  }
  get persistent() {
    return this.#inner.persistent === true;
  }
  layer(target, key) {
    return this.#inner.layer(target, key);
  }
  flushEvents() {
    this.#inner.flushEvents?.();
  }
  async #apply(target, directives, attempt, effect) {
    const index = this.#cursor.nextIndex(target);
    const directive = directiveAt(directives, index);
    this.#cursor.bump(target);
    if (directive === null) return effect(attempt);
    if (directive.delay_ms) await sleep(Math.max(0, Number(directive.delay_ms)));
    const outcome = resolve(directive);
    if (outcome === null) return effect(attempt);
    if (outcome === "crash") {
      this.#crash();
      throw new Error("unreachable: crash() must not return");
    }
    return outcome;
  }
}

function loadFaults(planPath) {
  let data;
  try {
    data = JSON.parse(readFileSync(planPath, "utf8"));
  } catch (err) {
    process.stderr.write(`keel ▸ KEEL_SIM_PLAN=${planPath} could not be read: ${err.message}\n`);
    process.exit(1);
  }
  const faults = data && typeof data === "object" ? data.faults : null;
  if (!faults || typeof faults !== "object") return {};
  const out = {};
  for (const [target, directives] of Object.entries(faults)) {
    if (Array.isArray(directives)) {
      out[target] = directives.filter((d) => d && typeof d === "object");
    }
  }
  return out;
}

/** Wrap `backend` for `keel sim <plan>` (`KEEL_SIM_PLAN=<planPath>`). */
export function installSim(backend, { planPath, env }) {
  const faults = loadFaults(planPath);
  const cursor = new Cursor(`${planPath}.cursor.json`);
  if (!isTruthy(env.KEEL_QUIET)) {
    process.stderr.write(`keel ▸ fault-simulating with ${planPath} — see docs/sim-format.md\n`);
  }
  return new SimBackend(backend, faults, cursor);
}
